src/front/wgsl/mod.rs

```diff
@@ -857,25 +858,41 @@ impl Parser {
                     }
                     lexer.expect(Token::Paren(')'))?;
                     if components.is_empty() {
-                        let last_component_inner = ctx.resolve_type(last_component)?;
-                        match (&ty.inner, last_component_inner) {
-                            (crate::TypeInner::Scalar { .. }, crate::TypeInner::Scalar { .. })
-                            | (crate::TypeInner::Matrix { .. }, crate::TypeInner::Matrix { .. })
-                            | (crate::TypeInner::Vector { .. }, crate::TypeInner::Vector { .. }) => {
-                                crate::Expression::As {
-                                    expr: last_component,
-                                    kind: kind.ok_or(Error::BadTypeCast(word))?,
-                                    convert: true,
-                                }
+                        // The issue here is that both `ctx.types[handle]` and `last_component_inner` borrow ctx (and one is mutable).
+                        // This isn't pretty.
+
+                        let last_component_ty = {
+                            let last_component_inner = ctx.resolve_type(last_component)?;
+                            match last_component_inner {
+                                crate::TypeInner::Scalar { .. } => Some(0),
+                                crate::TypeInner::Matrix { .. } => Some(1),
```

Comment:
why is `Matrix` associated with `1`?

Reply:
I just picked random numbers here, so we can check if the types are the same without running into borrow checking issues.